nifi-nar-bundles/nifi-gcp-bundle/nifi-gcp-processors/src/main/java/org/apache/nifi/processors/gcp/pubsub/publish/FlowFileResult.java

```diff
@@ -0,0 +1,143 @@
+        this.flowFile = flowFile;
+        this.attributes = new HashMap<>();
+        this.futures = futures;
+        this.successes = successes;
+        this.failures = failures;
+    }
+
+    /**
+     * After all in-flight messages have results, calculate appropriate {@link Relationship}.
+     */
+    public Relationship reconcile() {
+        while (futures.size() > (successes.size() + failures.size())) {
+            try {
+                ApiFutures.allAsList(futures).get();
+            } catch (InterruptedException | ExecutionException e) {
+                logger.error("Error while reconciling PublishGCPubSub send", e);
+            }
+        }
+        if (futures.size() == successes.size()) {
+            if (futures.size() == 1) {
+                attributes.put(PubSubAttributes.MESSAGE_ID_ATTRIBUTE, successes.iterator().next());
+            } else {
+                attributes.put(PubSubAttributes.RECORDS_ATTRIBUTE, Integer.toString(futures.size()));
+            }
+        }
+        return RelationshipMapper.toRelationship(failures);
+    }
+
+    public FlowFile getFlowFile() {
+        return flowFile;
+    }
+
+    public Map<String, String> getAttributes() {
+        return attributes;
+    }
+
+    /**
+     * Logic to derive an appropriate {@link Relationship} from the feedback provided by the client library.
+     * <p>
+     * Each {@link com.google.pubsub.v1.PubsubMessage} is associated with a {@link NiFiApiFutureCallback} at time of
+     * submission to the client library.  This callback allows the client library to convey information to the caller
+     * about the result of the (asynchronous) send.  If a send fails, an appropriate exception is conveyed, providing
+     * detail about the send failure; otherwise a message id (provided by the service) is supplied.
+     * <p>
+     * Types of exceptions might be classified into "retryable" (another send may be attempted) or non-retryable.
+     */
+    private static class RelationshipMapper {
+
+        private static Relationship toRelationship(final List<Throwable> failures) {
+            Relationship relationship = PublishGCPubSub.REL_SUCCESS;
+            boolean isRetry = false;
+            boolean isFailure = false;
+            for (final Throwable failure : failures) {
+                if (isRetryException(failure)) {
+                    isRetry = true;
+                } else {
+                    isFailure = true;
+                    break;
+                }
+            }
+            if (isFailure) {
+                relationship = PublishGCPubSub.REL_FAILURE;
+            } else if (isRetry) {
+                relationship = PublishGCPubSub.REL_RETRY;
+            }
+            return relationship;
+        }
+
+        /**
+         * Retryable exceptions indicate transient conditions; another send attempt might succeed.
+         */
+        private static final Collection<Class<? extends Throwable>> RETRY_EXCEPTIONS = Collections.singleton(
```

Comment:
This static member should be moved before the `toRelationship` method to follow standard ordering conventions.